Time: Implement BSD extensions to time

envp opened the pull request:
Provides:
- [X] `-f`: Format stats with an arbitrary printf-like `%<spec>` formatters 
- [X] `-t`: Format stats using the `tcsh(1)` format string
  - See: https://man.netbsd.org/time.1
- [X] `-c`: Write `time` stats in the `csh(1)` format string
  - See: https://man.netbsd.org/time.1
- [ ] `-l`: Dump the contents of the `rusage` struct (⚠️ Unimplemented as of this PR)
- [X] `-a`: If `-o` was provided, append to the target file instead of truncating
- [X] `-o`: Create, truncate a file to write `time` output to.

Part of: #53

envp committed:
Time: Implement BSD extensions to time

Provides implementations for:
- `-f`: Print rusage fields with a printf-like fmt spec
- `-t`: tcsh(1)-like output
- `-c`: csh(1)-like output
- `-l`: dump rusage contents

envp commented:
@GrayJack How do I get clap to not escape control characters such as "\t" in the CLI input? 

I was trying out the format strings with some control chars in them, but seems they are escaped by default, and I'm not sure how to turn that off for a particular option (and unescaping after-the-fact seems a rather odd thing to do as well)

```
$ cargo run --bin time -- -f "%Uu %Ss %E %P\t%X+%Dk %I+%Oio %Fpf+%Ww" sleep 1
0.76u 0.24s 1.01 0.99\t0+0k 0+85954560io 1115pf+0w
```

marcospb19 commented:
I took some time for trying understanding this.

It seems that "Escaping" is the default behavior, and treating `'\n'` and `'\t'` as special characters (line break and tab, as an example) will always require extra logic, that is, searching for `'\'` and checking the next character (unescaping is extra work).

However, some shells have their built-in versions of some coreutils with this feature, I assume that your shell is running a built-in version of `time` that does it.

@GrayJack correct me if I'm mistaken, but currently there is no util in this repo that unescapes arguments, not even `echo` (built-in version of `zsh` does).

`clap` does not support this operation.

GrayJack commented:
Sorry, I was looking in the cellphone app and by mistake I marked this pull request as ready for review.

I think @marcospb19 is correct. At least the version of clap we're using doesn't support unescaping arguments. Maybe the version 3.0 have this feature?

GrayJack commented:
CI is failing because of "if in const fn" in the clap app, inside a `bitflags!` macro.

Looks like clap set `bitflags` dependency to "1.0", but 1 on 1.3.0 a function was turned to const fn

envp committed:
Time: Unescape strings passed to `-f`

Format code

envp committed:
Time: Add documentation

envp commented:
Seems this is blocked on:

```
error[A001]: Potential segfault in the time crate
    ┌─ /github/workspace/Cargo.lock:106:1
    │
106 │ time 0.2.22 registry+https://github.com/rust-lang/crates.io-index
```

And this is `time`, the external command:

```
$ git grep '0.2.22'
coreutils_core/Cargo.toml:36:time = "= 0.2.22"
touch/Cargo.toml:18:time = "= 0.2.22"
```

I think it would make grepping the error messages easier if the `time` sub-crate was renamed to something else.

envp committed:
Time: cli: Add freebsd CLI options

envp committed:
Core: resource: Use `Duration` to represent time intervals

envp committed:
Time: Refactor internals to use `Duration``

- output: Introduce new `--human_readable` flag
- output: Split output formatting interface from the formatting
          strategy

envp committed:
Time: cli: Rename CLI parameters

Add more OSes from the BSD family

envp committed:
Time: Implement `-o`, `-a`

Adds the ability to make `time` write to a given
file, creating it if it doesn't exist (`-o`), or
optionally appending to it (`-a`)

envp committed:
Time: Fixup formatting

envp committed:
General: Bump `time` dependency to `0.2.23`

envp committed:
Core: Express nanoseconds as i32, seconds as i64

Also add an explanatory comment about why this is okay and
what it means if `time` panicked at that point

envp committed:
Time: cli: Remove `macos` freebsd group

It belongs with OpenBSD, which supports similar flags

envp commented:
It is really hard to implement the `-l` flag without some FFI.  I think it will be good to have that come through another PR.

I'll leave info here for follow-up in case me or someone else is interested at a later time: 

- Apple's time.c: https://opensource.apple.com/source/shell_cmds/shell_cmds-216.60.1/time/time.c.auto.html

The region of interest would be to look at:

```c
if (rusage_ret >= 0) {
	if (ruinfo.ri_instructions > 0) {
		fprintf(stderr, "%20" PRIu64 "  %s\n", ruinfo.ri_instructions,
			"instructions retired");
	}
	if (ruinfo.ri_cycles > 0) {
		fprintf(stderr, "%20" PRIu64 "  %s\n", ruinfo.ri_cycles,
			"cycles elapsed");
	}
	if (ruinfo.ri_lifetime_max_phys_footprint > 0) {
		fprintf(stderr, "%20" PRIu64 "  %s\n",
			ruinfo.ri_lifetime_max_phys_footprint,
			"peak memory footprint");
	}
}
``` 

This uses a `struct rusage_info_v4 ruinfo`, that is returned by `proc_pid_rusage(pid, RUSAGE_INFO_V4, (void **)&ruinfo)`. `time` should call this somehow to fill up some platform dependent / `Optional` fields in the `RUsage` struct.

envp commented:
@GrayJack This is ready for review, sorry for the commit spam :)

envp committed:
Time: flags: Make `destination`, `append` private

Valid combinations of these two are exposed through `get_output_stream`

envp commented:
ping!

Any reviewers open to take a look at this?

GrayJack commented:
Sorry for the late reply @envp , I'm getting used to my new job, but I'll review this PR either today or in the next weekend

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
+        .long("append-mode")
+        .short("a");
+
+    if cfg!(target_os = "netbsd") {
+        app.args(&[use_csh_fmt, format_string, dump_rusage, use_tcsh_fmt])
+    } else if cfg!(any(target_os = "freebsd", target_os = "dragonflybsd")) {
+        app.args(&[append_output, dump_rusage, human_readable, output_path])
+    } else if cfg!(any(target_os = "openbsd", target_os = "macos")) {
+        app.arg(dump_rusage)
+    } else {
+        app
+    }
```
Does those new features need to be only available in this logic? Or all those can be available on all platforms?

I'm asking because if those features are possible to be added to all OSes, it would be great

envp replied:
There's nothing platform specific about the individual flags (except for `-l`, which this PR doesn't implement anyway). They're only gated right now because the original binaries have varying feature-set across different platforms. I think it would be better to merge them anyway since it makes it easier to test the binary

GrayJack commented on time/src/main.rs:
```diff
@@ -14,6 +14,14 @@ fn main() {
+        Err(err) => Err(err),
+    };
+
+    written.err().map(subprocess::exit_with_msg);
```
On cases that a function returns unit and never is more idiomatic to use `if let`, even though it would use more lines

GrayJack commented on time/src/output.rs:
````diff
@@ -1,32 +1,261 @@
+    /// ```
+    /// %Uu %Ss %E %P\t%X+%Dk %I+%Oio %Fpf+%Ww
+    /// ```
+    TCSH,
````
Those two names as not idiomatic for enum variants. I think `Csh` and `Tcsh` (or `TCsh`) would be more idiomatic

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
+        .short("h");
+
+    let output_path = Arg::with_name("output_file")
+        .takes_value(true)
```
`Arg` uses the name passed to `with_name` as the default name for the value taken in the help message. We've been using all caps for the value name taken by options. You can use `value_name("NAME_OF_THE_VALUE")` to set the value name in the help message, also `value_name` implicitly sets `takes_value` to true, so you wont need the `takes_value` anymore

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
+        .short("c");
+    let format_string = Arg::with_name("format_string")
+        .conflicts_with_all(&["posix", "use_csh_fmt"])
+        .takes_value(true)
```
`Arg` uses the name passed to `with_name` as the default name for the value taken in the help message. We've been using all caps for the value name taken by options. You can use `value_name("NAME_OF_THE_VALUE")` to set the value name in the help message, also `value_name` implicitly sets `takes_value` to true, so you wont need the `takes_value` anymore

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
+        .takes_value(true)
+        .help("Write the output to file instead of stderr.\
+              If file exists and the -a flag is not specified,\
+              the file will be overwritten.")
```
It would be better if for `help` only the first sentence is displayed, and the entire message with the `long_help`

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
+                Ok(())
+            }
+        })
+        .help(FMT_ARG_HELP)
```
It would be better if for `help` only the first sentence is displayed, and the entire message with the `long_help`

GrayJack commented on time/src/cli.rs:
```diff
@@ -19,14 +65,70 @@ pub(crate) fn create_app<'a, 'b>() -> App<'a, 'b> {
-                .long("posix")
-                .short("p"),
-        )
+        .arg(Arg::with_name("posix").help(POSIX_FMT_HELP).long("posix").short("p"));
```
It would be better if for `help` only the first sentence is displayed, and the entire message with the `long_help`

GrayJack reviewed (commented):
Thanks for another contribution @envp!! 🎉 

I just finished reviewing your code! The most important comments for me in on cli.rs, the other one are mostly small nitpicks.

One thing I noticed is that if the code fails to write to the output, the error message doesn't have the name of the program and the name of the output, just the error message. I don't know where is the code of this error printing, I'll need more time to find it, but if you know where it is, can you point it out to me? No need to fix in this PR, I think the PR is in a great state and this specific problem should be part of another PR

I also wanted to ask, in the function `custom_formatter` there is some `expect`, this is intentional, correctly?

envp commented:
> Sorry for the late reply @envp , I'm getting used to my new job, but I'll review this PR either today or in the next weekend

No worries :)

I understand it can be hard to make time with (especially new) full-time responsibilities!

envp committed:
Time: cli: Merge cross-platform features

Improve help text: Use `long_help` for multiline help text

envp committed:
Time: Rename `FormatterKind` variants

Use CamelCase instead of ALLCAPS

envp committed:
Time: main: Rewrite code for failing on error

envp committed:
Time: Version bump

envp commented:
> One thing I noticed is that if the code fails to write to the output, the error message doesn't have the name of the program and the name of the output, just the error message. I don't know where is the code of this error printing, I'll need more time to find it, but if you know where it is, can you point it out to me? No need to fix in this PR, I think the PR is in a great state and this specific problem should be part of another PR

Sure! I'd be happy to fix it in another PR :). I wrote the `.expect` to silence `rustc` about unused results, but now that they're there, I may as well improve them enough that they're good assertions. 

> I also wanted to ask, in the function `custom_formatter` there is some `expect`, this is intentional, correctly?

Yep, I'm not sure how to handle those exactly. Would it be better to wrap the errors in the individual formatters, instead of failing eagerly? (All the expects are at locations where the error is unrecoverable, i.e. user asked for output but we failed to write to the format buffer for whatever reason.)

GrayJack reviewed (approved):
Looks good to me

bors r+

GrayJack commented:
bors r+

GrayJack commented:
bors r+

If it doesn't work this time, I'll consider a manual merge

envp commented:
> Timed out.

:(
